### PR TITLE
4. bulk job manager

### PR DESCRIPTION
**Pull request by msvinaykumar**
## Description

This is JOB manager class used to handle jobs created using Bulk Api service .

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [ ] Requires DB changes

## How has this been tested?

Please describe the tests that were run to verify your changes and steps to reproduce. Please specify any test configuration required. 

- [ ] New Test X
- [ ] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: 

## Checklist :dart:

- [ ] Followed coding guidelines
- [ ] Comments added
- [ ] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information

Include any additional information such as links, test results, screenshots here

### CODE REVIEW

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                                            createExperimentAPIObject.setMode("monitor");
+                                            createExperimentAPIObject.setTargetCluster("local");
+                                            createExperimentAPIObject.setApiVersion("v2.0");
+                                            String experiment_name = "prometheus-1" + "-" + dsc.getDataSourceClusterName() + "-" + namespace.getDataSourceNamespaceName()
+                                                    + "-" + dsw.getDataSourceWorkloadName() + "(" + dsw.getDataSourceWorkloadType() + ")" + "-" + dc.getDataSourceContainerName();
+                                            createExperimentAPIObject.setExperimentName(experiment_name);
+                                            createExperimentAPIObject.setDatasource("prometheus-1");
+                                            createExperimentAPIObject.setClusterName(dsc.getDataSourceClusterName());
+                                            createExperimentAPIObject.setPerformanceProfile("resource-optimization-openshift");
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java:126]: This should be `resource-optimization-local-monitoring` - https://github.com/kruize/autotune/blob/master/manifests/autotune/performance-profiles/resource_optimization_local_monitoring.json

Comment (dinogun) [inline src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java:126]: Pls use constants for all the strings

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                    if (includeLabelsBuilder.length() > 0) {
+                        includeLabelsBuilder.setLength(includeLabelsBuilder.length() - 1);
+                    }
+                    LOGGER.info("Include Labels: " + includeLabelsBuilder.toString());
+                    uniqueKey = includeLabelsBuilder.toString();
+                }
+            }
+            DataSourceMetadataInfo metadataInfo = null;
+            DataSourceManager dataSourceManager = new DataSourceManager();
+            DataSourceInfo datasource = CommonUtils.getDataSourceInfo("prometheus-1");
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java:78]: If you are making any assumptions on the datasource, move it all to a constants file and add a TODO section to use actual values in the input JSON and not make any assumptions

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                            }
+                        }
+                    }
+                }
+                jobStatusMap.get(jobID).setStatus("INPROGRESS");
+                jobStatusMap.get(jobID).getData().getRecommendations().getData().setInqueue(recommendationsRequiredExperiments);
+                jobStatusMap.get(jobID).getData().getRecommendations().setTotalCount(recommendationsRequiredExperiments.size());
+
+            }
+            ExecutorService executor = Executors.newFixedThreadPool(3);
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java:172]: Why hardcoding to 3 threads? How will this affect scalability

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                LocalDateTime localDateTime = LocalDateTime.parse(intervalEndTimeStr, DateTimeFormatter.ofPattern(KruizeConstants.DateFormats.STANDARD_JSON_DATE_FORMAT));
+                interval_end_time_epoc = localDateTime.toEpochSecond(ZoneOffset.UTC);
+                Timestamp interval_end_time = Timestamp.from(localDateTime.toInstant(ZoneOffset.UTC));
+                localDateTime = LocalDateTime.parse(intervalStartTimeStr, DateTimeFormatter.ofPattern(KruizeConstants.DateFormats.STANDARD_JSON_DATE_FORMAT));
+                interval_start_time_epoc = localDateTime.toEpochSecond(ZoneOffset.UTC);
+                Timestamp interval_start_time = Timestamp.from(localDateTime.toInstant(ZoneOffset.UTC));
+
+                int steps = 15 * KruizeConstants.TimeConv.NO_OF_SECONDS_PER_MINUTE;
+                //Get metaData
+                //example metadataInfo = dataSourceManager.importMetadataFromDataSource(datasource, uniqueKey, interval_start_time_epoc, interval_end_time_epoc, steps);
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java:96]: Looks like redundant code, please remove them

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                            if (dataSourceWorkloadHashMap != null) {
+                                for (DataSourceWorkload dsw : dataSourceWorkloadHashMap.values()) {
+                                    HashMap<String, DataSourceContainer> dataSourceContainerHashMap = dsw.getDataSourceContainerHashMap();
+                                    if (dataSourceContainerHashMap != null) {
+                                        for (DataSourceContainer dc : dataSourceContainerHashMap.values()) {
+                                            CreateExperimentAPIObject createExperimentAPIObject = new CreateExperimentAPIObject();
+                                            createExperimentAPIObject.setMode("monitor");
+                                            createExperimentAPIObject.setTargetCluster("local");
+                                            createExperimentAPIObject.setApiVersion("v2.0");
+                                            String experiment_name = "prometheus-1" + "-" + dsc.getDataSourceClusterName() + "-" + namespace.getDataSourceNamespaceName()
```
[the remark this answers is not in the document]
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: done

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java (hunk 1 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: done

Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: done

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java (hunk 2 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: done

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java (hunk 3 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: Now I made this configurable vai env .  Adjust the thread pool size based on the kruize pod count 
at most three threads will be allowed to execute concurrently.

in future we can use inline with Keda plugin and have  dynamic number of threads to scale based on demand.

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java (hunk 4 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: The commented code will be uncommented in the next PR; it was only used as a placeholder for now

**Comment by msvinaykumar** [comment]
checking

File: src/main/java/com/autotune/analyzer/kruizeObject/CreateExperimentConfigBean.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
+package com.autotune.analyzer.kruizeObject;
+
+public class CreateExperimentConfigBean {
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/kruizeObject/CreateExperimentConfigBean.java:3]: Why is this needed? Dont we already have Experiment related classes? Please add a javadoc with sufficient explanation

Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/kruizeObject/CreateExperimentConfigBean.java:3]: This object is placeholder for create Experiment template which holds default values which is set in 

https://github.com/kruize/autotune/blob/ffe14a9eec97a359bc877cd8759021f763b9ead1/src/main/java/com/autotune/utils/KruizeConstants.java#L780C1-L788C70

Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/kruizeObject/CreateExperimentConfigBean.java:21]: Add javadoc with explination

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                Map<String, String> includeLabels = this.bulkInput.getFilter().getInclude().getLabels();
+                if (includeLabels != null && !includeLabels.isEmpty()) {
+                    includeLabels.forEach((key, value) ->
+                            includeLabelsBuilder.append(key).append("=").append("\"" + value + "\"").append(",")
+                    );
+                    // Remove trailing comma
+                    if (includeLabelsBuilder.length() > 0) {
+                        includeLabelsBuilder.setLength(includeLabelsBuilder.length() - 1);
+                    }
+                    LOGGER.info("Include Labels: " + includeLabelsBuilder.toString());
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java:116]: Looks like all the LOGGER.info should be debug instead. Please go through all of them and retain only what is necessary as info

File: src/main/java/com/autotune/utils/KruizeConstants.java [hunk 8]
```diff
@@ -674,6 +676,8 @@ public static final class KRUIZE_CONFIG_ENV_NAME {
         public static final String CLOUDWATCH_LOGS_LOG_LEVEL = "logging_cloudwatch_logLevel";
         public static final String LOCAL = "local";
         public static final String LOG_HTTP_REQ_RESP = "logAllHttpReqAndResp";
+        public static final String RECOMMENDATIONS_URL = "recommendationsURL";
```
Comment (dinogun) [inline src/main/java/com/autotune/utils/KruizeConstants.java:679]: Why do we need `recommendationsURL`?

Comment (msvinaykumar) [reply src/main/java/com/autotune/utils/KruizeConstants.java:679]: this is required to specifiy where krizue recommendation pod are located . if it is localhost then we need to specify 
service_name.namespace.svc.local.net in manifest which @saad create pr for same

File: src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java (hunk 7 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/workerimpl/BulkJobManager.java]: done

File: src/main/java/com/autotune/utils/KruizeConstants.java (hunk 8 above)
Comment (khansaad) [reply src/main/java/com/autotune/utils/KruizeConstants.java:679]: PR [1335](https://github.com/kruize/autotune/pull/1335) has the manifest changes for the same

**Review by dinogun — APPROVED** [review]
LGTM
